File: Leetcode/python/Array Nesting.py
class Solution:
    def arrayNesting(self, nums):
        mx = 0
        nest = set()
        for i in range(len(nums)):
            if nums[i] not in nest:
                iset = set()
                init = i

                index = i
                while 1:
                    if index in iset:
                        if index == init:
                            mx = max(len(iset), mx)
                        break

                    else:
                        iset.add(index)
                        index = nums[index]

                nest = nest | iset

        return mx



s = Solution()
nums = [5,4,0,3,1,6,2]
print(s.arrayNesting(nums))

File: Leetcode/C++/Array Nesting.cpp
class Solution {
public:
    int arrayNesting(vector<int>& nums) {
    	int mx = 0;
    	int l = nums.size();
    	set<int> nest;
        for(int i=0; i<l; i++){
        	if(nest.find(nums[i]) != nest.end()){

        		int init = i;
        		int index = i;
        		set<int> iset;

        		while 1{
        			if(iset.find(index) != iset.end()){
        				if(index == init)
        					mx = max(mx, iset.size());
        				break;
        			}
        			else{
        				nest.insert(index);
        				index = nums[index];
        			}
        		}
        		set_intersection(nest.begin(), nest.end(), iset.begin, iset.end(), inserter(nest, nest.end()))
        	}
        }
        return mx;
    }
};